Assignment_05.rb

```diff
@@ -0,0 +1,133 @@
+=begin
+Description
+
+Assignment:
+
+1.
+sample input structure: (suppose this is the list of tasks & related time log data for an employee for a particular day from morning 9:30am to evening 6:30pm )
+
+[{ id: <task1 id>, name: <task1_name>, time_logs: [[<start_time>, <end_time>], [<start_time>, <end_time>]], ...}, <task2>, <task3>, ...]
+
+eg:
+(sample valid input)
+[
+    { id: 1, name: "task1", time_logs: [[DateTime.parse("2023-01-01 9:30:00"), DateTime.parse("2023-01-01 10:30:00")],
+                                        [DateTime.parse("2023-01-01 12:00:00"), DateTime.parse("2023-01-01 13:00:00")]] },
+    { id: 2, name: "task2", time_logs: [[DateTime.parse("2023-01-01 10:30:00"), DateTime.parse("2023-01-01 11:30:00")],
+                                        [DateTime.parse("2023-01-01 13:00:00"), DateTime.parse("2023-01-01 14:30:00")]] },
+    { id: 3, name: "task3", time_logs: [[DateTime.parse("2023-01-01 11:30:00"), DateTime.parse("2023-01-01 12:00:00")],
+                                        [DateTime.parse("2023-01-01 14:30:00"), DateTime.parse("2023-01-01 15:30:00")]] },
+    { id: 4, name: "task4", time_logs: [[DateTime.parse("2023-01-01 15:30:00"), DateTime.parse("2023-01-01 16:30:00")]] },
+    { id: 5, name: "task5", time_logs: [[DateTime.parse("2023-01-01 16:30:00"), DateTime.parse("2023-01-01 17:30:00")]] },
+    { id: 6, name: "task6", time_logs: [[DateTime.parse("2023-01-01 17:30:00"), DateTime.parse("2023-01-01 18:30:00")]] },
+  ]
+
+
+expected output format:
+
+  {
+    categories: ["task1", "task2", "task3", "task4", "task5", "task6"],
+    total_time: [<total_spent_time_in_minutes_for_task1>,  .. <total_spent_time_in_minutes_for_task6>],
+    total_time: [<(total_spent_time_in_minutes_for_task_1 / (24 * 60)) * 100>, .. <(total_spent_time_in_minutes_for_task_6 / (24 * 60)) * 100>]
+  }
+
+  
+1. Define required class & methods for it
+2. split down the main method into multiple methods meaningfully
+3. validate input & handle exception for invalid input
+4. write unit test cases for this class functionalities using rspec
+=end
+
+require 'date'
+
+class Task
+  attr_accessor :id, :name, :time_logs
+
+  def initialize(id="deafult", name="deafult", time_logs=[])
+    @id = id
+    @name = name
+    @time_logs = time_logs
+  end
+
+  # This method returns the total time used for the task
+  def total_time_in_minute
+    time_logs.sum { |time_log| get_time_duration_in_minute(time_log[0], time_log[1]) }.round()
+  end
+
+  # This method takes total time used for a task and returns the in percentage form for a day
+  def occupation_percecntage_day_wise(total_time_in_minute)
+    (total_time_in_minute.to_f / (24 * 60) * 100).round(2)
+  end
+
+  # This method validate the the time logs
+  def validate_time_logs
```

Comment:
should be under private access

Reply:
I am using the validate_time_logs outside the Task class, that is why it is public. I am working on to make it private.